pkg/probe/probe.go

```diff
@@ -80,3 +83,58 @@ func HaveIPv6Support() bool {
 	unix.Close(fd)
 	return true
 }
+
+// HaveSourceOuterIPSupport tests whether the kernel support setting the outer
+// source IP address via the bpf_skb_set_tunnel_key BPF helper. We can't rely
+// on the verifier to reject a program using the new support because the
+// verifier just accepts any argument size for that helper; non-supported
+// fields will simply not be used. Instead, we set the outer source IP and
+// retrieve it with bpf_skb_get_tunnel_key right after. If the retrieved value
+// equals the value set, we have a confirmation the kernel supports it.
+func HaveOuterSourceIPSupport() (bool, error) {
```

Comment:
I just noticed we have two packages for feature probing. Could we add this function to `pkg/datapath/linux/probes` instead to keep adding feature probes to only one of them? I'd offer to take care of migrating the rest of the probes existing in `pkg/probe`.

With regards to the API, I would like to suggest to only return an error like all the other feature probing APIs in `pkg/datapath/linux/probes` with following return semantics:
* err == nil: feature is supported
* errors.Is(err, probes.ErrNotSupported): feature is not supported

If we encounter unexpected errors while executing the feature probe itself we fatal in the probe function itself. This would slightly simplify the calling code as well:
```golang
if err := HaveOuterSourceIPSupport(); err != nil {
    log.Fatal(...)
}
```

Not sure if we'd ever have to call this feature probe in a different place than this, but if we do it would be nice to cache the returned error with a global var and only run the actual probe once with `sync.Once`. 

Lastly, could we add a small test for this that checks if it succeeds on kernels where this is supported? We just landed `testutils.SkipOnOldKernel()` for tests like that.

Reply:
> If we encounter unexpected errors while executing the feature probe itself we fatal in the probe function itself. This would slightly simplify the calling code as well:

I'm not exactly sure what you meant by that. That's already the case. We fatal in the caller if it fails. We have two cases because I want to specialize the message depending on whether the feature is missing vs. probing failed.

> Not sure if we'd ever have to call this feature probe in a different place than this, but if we do it would be nice to cache the returned error with a global var and only run the actual probe once with sync.Once.

I'd be surprised if we ever need this probe somewhere else. Setting our own source IP for the tunnel is very specific to the high-scale IPcache feature. Also, a bit fearful that if I do that, the next reviewer is then going to ask me to avoid global variables :laughing: So skipped this for now.

<br>

I moved the probe, reworked its return semantics, and added a test for it.

Reply:
> That's already the case. We fatal in the caller if it fails. We have two cases because I want to specialize the message depending on whether the feature is missing vs. probing failed.

I was trying to say that I would move the fataling in case of probe failures into the probe directly to stay consistent across all probing APIs. So like [HaveProgramType](https://github.com/cilium/cilium/blob/3e805306fc4125fca3398397acdf11a13f1a9600/pkg/datapath/linux/probes/probes.go#L358), if an unexpected error occurs, we call log.Fatal directly in the probe.

Everything else LGTM! :rocket: